test: ensure test vars are not used in tf plan

AdosH1 opened the pull request:
Added test to ensure variables in a tests folder are not applied upon a `tf plan` command.
Related to work discussed in [#2106](https://github.com/opentofu/opentofu/pull/2106)
Stemming from issue #1367 
## Target Release

N/A

## Checklist

<!-- Please check of ALL items in this list for all PRs: -->

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [x] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

<!-- If your PR contains Go code, please make sure you check off all items on this list: --> 

- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

ollevche commented on internal/command/plan_test.go:
```diff
@@ -220,6 +220,44 @@ func TestPlan_noState(t *testing.T) {
+
+	plan := testReadPlan(t, outPath)
+	ctyTestVar := plan.VariableValues["testVar"]
+	psuedo, err := ctyTestVar.Decode(cty.DynamicPseudoType)
```
This is fine - variables are wrapped as `cty.DynamicPseudoType`, but after decoding it is okay to use it as a `cty.String` in this case.

AdosH1 replied:
Thanks for the explanation, its weird to me `cty`'s own encode/decode functions don't quite work together as well as I'd hope

ollevche commented on internal/command/plan_test.go:
```diff
@@ -220,6 +220,44 @@ func TestPlan_noState(t *testing.T) {
+
+	plan := testReadPlan(t, outPath)
+	ctyTestVar := plan.VariableValues["testVar"]
+	psuedo, err := ctyTestVar.Decode(cty.DynamicPseudoType)
```
`psuedo` should be renamed like `decodedExpectedVar` to be more readable.

AdosH1 replied:
agreed, updated :pray:

ollevche commented on internal/command/plan_test.go:
```diff
@@ -220,6 +220,44 @@ func TestPlan_noState(t *testing.T) {
+		t.Fatalf("Unable to decode plan variables, error: %v", err)
+	}
+
+	result := psuedo.GoString()
```
You can use `psuedo.AsString()` here because it is fine to assume the decoded var is of `cty.String` type in that particular case.

ollevche commented on internal/command/plan_test.go:
```diff
@@ -220,6 +220,44 @@ func TestPlan_noState(t *testing.T) {
+	}
+
+	result := psuedo.GoString()
+	if !strings.Contains(result, "ValueFROMmain/tfvars") {
```
Let's move the expected value as a separate constant to reduce duplication.

AdosH1 replied:
thanks, updated

ollevche commented on internal/command/plan_test.go:
```diff
@@ -220,6 +220,44 @@ func TestPlan_noState(t *testing.T) {
+		"-out", outPath,
+	}
+	code := c.Run(args)
+	output := done(t)
```
As an alternative approach, you can check the output of the plan command instead of the plan file. You would want to check if `test_instance.testRes.ami` equals to expected value (`ValueFROMmain/tfvars`) in the output. This alternative approach could simplify the test a bit.

AdosH1 replied:
I've replaced this with a simple `contains` in the output string, as I don't want to dictate any formatting requirements in the output
Given the test case, there won't be anything else that would assume the value `valueFROMmain/tfvars`so I think this is okay - let me know if you think we need to be more specific

ollevche replied:
I don't think it is required, but I would use a more specific line to search for. Also, I would double-check if the output contains `ValueFROMtests/tfvars` just to be sure it doesn't.

AdosH1 replied:
Updated to assert for the line `ami = "ValueFROMmain/tfvars"`
Added a check to ensure `ValueFROMtests/tfvars` doesn't exist in output

ollevche reviewed (commented):
Thanks for the contribution @AdosH1! The test is fine already and I left a few suggestions to improve.

ollevche commented:
Also, @AdosH1 you have to add a DCO sign to all of your commits: [details](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md#signing-off-your-commits).

AdosH1 commented:
Thank you for your comments @ollevche, I appreciate your support!
I've updated the PR based on your comments and simplified the assertion, let me know if there's anything else you think might need to be improved :pray:

ollevche commented:
Hey @AdosH1 I left a minor improvement comment. Otherwise, the PR looks good!

ollevche commented on internal/command/plan_test.go:
```diff
@@ -221,6 +221,41 @@ func TestPlan_noState(t *testing.T) {
+	}
+
+	// Verify values defined in the 'tests' folder are not used during a plan action
+	expectedValue := "ValueFROMmain/tfvars"
+	expectedResult := fmt.Sprintf("ami = \"%s\"", expectedValue)
```
nit:

```suggestion
	expectedResult := `ami = "ValueFROMmain/tfvars"`
```

ollevche commented on internal/command/plan_test.go:
```diff
@@ -221,6 +221,41 @@ func TestPlan_noState(t *testing.T) {
+		},
+	}
+
+	args := []string{}
+	code := c.Run(args)
```
nit:

```suggestion
	code := c.Run([]string{})
```

AdosH1 commented:
@ollevche Thanks for the suggestions, committed & ensured tests are passing
